src/base/static/components/templates/map.tsx

```diff
@@ -84,11 +79,81 @@ const SpotlightMask = styled("div")({
   zIndex: 8,
 });
 
-class MapTemplate extends Component {
-  mapContainerRef = createRef();
-  addPlaceButtonRef = createRef();
+const statePropTypes = {
+  datasetsConfig: datasetsConfigPropType.isRequired,
+  hasAddPlacePermission: PropTypes.bool.isRequired,
+  hasGroupAbilitiesInDatasets: PropTypes.func.isRequired,
+  initialMapViewport: mapViewportPropType.isRequired,
+  isAddPlaceButtonVisible: PropTypes.bool.isRequired,
+  isContentPanelVisible: PropTypes.bool.isRequired,
+  isGeocodeAddressBarEnabled: PropTypes.bool.isRequired,
+  isLeftSidebarExpanded: PropTypes.bool.isRequired,
+  isRightSidebarEnabled: PropTypes.bool.isRequired,
+  isRightSidebarVisible: PropTypes.bool.isRequired,
+  isSpotlightMaskVisible: PropTypes.bool.isRequired,
+  mapSourceNames: PropTypes.arrayOf(PropTypes.string).isRequired,
+  layout: PropTypes.string.isRequired,
+  mapConfig: mapConfigPropType.isRequired,
+  navBarConfig: navBarConfigPropType.isRequired,
+  placeConfig: placeConfigPropType.isRequired,
+  placeExists: PropTypes.func.isRequired,
+};
+
+const dispatchPropTypes = {
+  createFeaturesInGeoJSONSource: PropTypes.func.isRequired,
+  loadPlaceAndSetIgnoreFlag: PropTypes.func.isRequired,
+  updateUIVisibility: PropTypes.func.isRequired,
+  updateActivePage: PropTypes.func.isRequired,
+  updateContentPanelComponent: PropTypes.func.isRequired,
+  updateFocusedPlaceId: PropTypes.func.isRequired,
+  updateEditModeToggled: PropTypes.func.isRequired,
+  updateScrollToResponseId: PropTypes.func.isRequired,
+};
+
+type StateProps = PropTypes.InferProps<typeof statePropTypes>;
+type DispatchProps = PropTypes.InferProps<typeof dispatchPropTypes>;
+interface MapTemplateOwnProps {
+  uiConfiguration: string;
+  isStartPageViewed?: boolean;
+  onViewStartPage?: () => void;
+  languageCode: string;
+  params: {
+    datasetClientSlug?: string;
+    responseId?: string;
+    pageSlug?: string;
+    placeId?: string;
+    zoom?: string;
+    lat?: string;
+    lng?: string;
+  };
+}
+interface State {
+  mapContainerHeightDeclaration: string;
+  mapContainerWidthDeclaration: string;
+  mapViewport: {
+    minZoom: number;
+    maxZoom: number;
+    bearing?: number | undefined | null;
```

Comment:
I'm not sure about this typing-- I think we need to include `undefined` and `null` because the `INITIAL_STATE` defined in the `map` duck doesn't define values for `bearing`, `zoom`, `latitude`, or `longitude` (since we don't want to introduce an arbitrary initial map position or zoom).

Reply:
I think I'd rather set the map state to initial numerical values, which seems like it would be cleaner.

The idea behind `initialMapViewport` was to keep track of both the default map viewport as set in the config, but also the viewport when navigating away from the `MapTemplate`. That way, the last viewport can be restored when returning to the `MapTemplate`.

Though we could also keep this state in `App`'s component state, and update it via a callback from `MapTemplate` when it unmounts.

Reply:
>Though we could also keep this state in App's component state, and update it via a callback from MapTemplate when it unmounts.

Seems like a reasonable intermediate solution. It seems like a more thorough solution, however, it to use the Context API: https://reactjs.org/docs/context.html

That should give us the persistence that global state provides, without the performance drawbacks that Redux entails, while also avoiding the prop-drilling that component state entails as well.